encoding/codecv2/codecv2.go

```diff
@@ -0,0 +1,507 @@
+	}
+
+	return chunkBytes
+}
+
+// Hash computes the hash of the DAChunk data.
+func (c *DAChunk) Hash() (common.Hash, error) {
+	var dataBytes []byte
+
+	// concatenate block contexts
+	for _, block := range c.Blocks {
+		encodedBlock := block.Encode()
+		// only the first 58 bytes are used in the hashing process
+		dataBytes = append(dataBytes, encodedBlock[:58]...)
+	}
+
+	// concatenate l1 tx hashes
+	for _, blockTxs := range c.Transactions {
+		for _, txData := range blockTxs {
+			if txData.Type == types.L1MessageTxType {
+				txHash := strings.TrimPrefix(txData.TxHash, "0x")
+				hashBytes, err := hex.DecodeString(txHash)
+				if err != nil {
+					return common.Hash{}, err
+				}
+				if len(hashBytes) != 32 {
+					return common.Hash{}, fmt.Errorf("unexpected hash: %s", txData.TxHash)
+				}
+				dataBytes = append(dataBytes, hashBytes...)
+			}
+		}
+	}
+
+	hash := crypto.Keccak256Hash(dataBytes)
+	return hash, nil
+}
+
+// NewDABatch creates a DABatch from the provided encoding.Batch.
+func NewDABatch(batch *encoding.Batch) (*DABatch, error) {
+	// this encoding can only support a fixed number of chunks per batch
+	if len(batch.Chunks) > MaxNumChunks {
+		return nil, fmt.Errorf("too many chunks in batch")
+	}
+
+	if len(batch.Chunks) == 0 {
+		return nil, fmt.Errorf("too few chunks in batch")
+	}
+
+	// batch data hash
+	dataHash, err := ComputeBatchDataHash(batch.Chunks, batch.TotalL1MessagePoppedBefore)
+	if err != nil {
+		return nil, err
+	}
+
+	// skipped L1 messages bitmap
+	bitmapBytes, totalL1MessagePoppedAfter, err := encoding.ConstructSkippedBitmap(batch.Index, batch.Chunks, batch.TotalL1MessagePoppedBefore)
+	if err != nil {
+		return nil, err
+	}
+
+	// blob payload
+	blob, blobVersionedHash, z, err := constructBlobPayload(batch.Chunks)
+	if err != nil {
+		return nil, err
+	}
+
+	daBatch := DABatch{
+		Version:                CodecV2Version,
+		BatchIndex:             batch.Index,
+		L1MessagePopped:        totalL1MessagePoppedAfter - batch.TotalL1MessagePoppedBefore,
+		TotalL1MessagePopped:   totalL1MessagePoppedAfter,
+		DataHash:               dataHash,
+		BlobVersionedHash:      blobVersionedHash,
+		ParentBatchHash:        batch.ParentBatchHash,
+		SkippedL1MessageBitmap: bitmapBytes,
+		blob:                   blob,
+		z:                      z,
+	}
+
+	return &daBatch, nil
+}
+
+// ComputeBatchDataHash computes the data hash of the batch.
+// Note: The batch hash and batch data hash are two different hashes,
+// the former is used for identifying a badge in the contracts,
+// the latter is used in the public input to the provers.
+func ComputeBatchDataHash(chunks []*encoding.Chunk, totalL1MessagePoppedBefore uint64) (common.Hash, error) {
+	return codecv1.ComputeBatchDataHash(chunks, totalL1MessagePoppedBefore)
+}
+
+// constructBlobPayload constructs the 4844 blob payload.
+func constructBlobPayload(chunks []*encoding.Chunk) (*kzg4844.Blob, common.Hash, *kzg4844.Point, error) {
```

Comment:
I think it's OK not to add a receiver to pure computing functions, because it can be used easier without initializing a receiver, and also it can reuse codecv1's implementations.

Reply:
if you prefer this function to a common method, maybe put it into `utils` or `common`